apps/passport-server/src/services/poapService.ts

```diff
@@ -57,58 +63,131 @@ export class PoapService {
     this.rollbarService = rollbarService;
   }
 
+  /**
+   * This helper function checks that a serialized PCD satisfies
+   * the following properties:
+   *   1. The type of the PCD is ZKEdDSAEventTicketPCD
+   *   2. The signer of the PCD's claim matches this server's EdDSA public key
+   *   3. The proof of the PCD is valid when checked by the verify() function
+   * If the serialized PCD satisfies these three properties, the deserialized
+   * ZKEdDSAEventTicketPCD is returned.
+   */
+  private async validateSignedZKEdDSAEventTicketPCD(
+    serializedPCD: string
+  ): Promise<ZKEdDSAEventTicketPCD> {
+    logger(
+      "[POAP] checking that PCD type is ZKEdDSAEventTicketPCD",
+      serializedPCD
+    );
+    const parsed = JSON.parse(serializedPCD) as SerializedPCD;
+    if (parsed.type !== ZKEdDSAEventTicketPCDPackage.name) {
+      throw new Error("proof must be ZKEdDSAEventTicketPCD type");
+    }
+
+    const pcd = await ZKEdDSAEventTicketPCDPackage.deserialize(parsed.pcd);
+
+    logger(
+      `[POAP] checking that signer of ticket ${pcd.claim.partialTicket.ticketId} is passport-server`
+    );
+    if (!process.env.SERVER_EDDSA_PRIVATE_KEY)
+      throw new Error(`missing server eddsa private key .env value`);
+
+    const TICKETING_PUBKEY = await getEdDSAPublicKey(
+      process.env.SERVER_EDDSA_PRIVATE_KEY
+    );
+
+    const signerMatch =
+      pcd.claim.signer[0] === TICKETING_PUBKEY[0] &&
+      pcd.claim.signer[1] === TICKETING_PUBKEY[1];
+
+    if (!signerMatch) {
+      throw new Error("signer of PCD is invalid");
+    }
+
+    logger("[POAP] verifying PCD proof and claim", pcd);
+    if (!(await ZKEdDSAEventTicketPCDPackage.verify(pcd))) {
+      throw new Error("pcd invalid");
+    }
+
+    return pcd;
+  }
+
   /**
    * Validates that a serialized ZKEdDSAEventTicketPCD is a valid
-   * Devconnect Cowork ticket that has been checked in, and returns
-   * the ID of that ticket.
+   * Zuzalu 2023 Ticket and returns the ID of that ticket.
    *
    * This function throws an error in the case that the PCD is not
    * valid; for example, here are a few invalid cases
    *  1. Wrong PCD type
    *  2. Wrong EdDSA public key
    *  3. PCD proof is invalid
-   *  4. Ticket does not exist
-   *  5. Ticket has not been checked in
-   *  6. Event of ticket is not Cowork space
-   *  7. Invalid product for claiming a poap, e.g. EF Towel
+   *  4. Event of ticket is not Zuzalu 2023
```

Comment:
Just curious but is this intended to be just for Zuzalu, or Zuconnect as well? Any reason to not give Zuconnect attendees one? Will that be the next request from their end?

Reply:
Just Zuzalu. ZuConnect was not proposed, and it requires another endpoint from our standpoint, so I elected not to do it for now.